### PR TITLE
Distribute DMN resource deletions

### LINKED ISSUE #13204: Distribute decision definition deletion

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

With the (almost) completion of #11456 we are now able to use this upon deleting a resource. As of this time only DMN resource deletion is implemented. This PR does some refactorings to this code and adds the distribution and acknowledgements.

I recommend reviewing this per commit.

## Related issues

closes #13204 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionProcessor.java
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionProcessor.java:68]: 🔧 In addition to the previous point, it's not clear what should happen here when `deleteResources` fails (it's highly unlikely and it really shouldn't, but it's possible). 

💭 Note that this is not clear anymore, now that the flow control is pushed down into that method and it did not have a name indicating that it could fail.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionMultiPartitionTest.java
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionMultiPartitionTest.java:60]: 🔧 As discussed this limit is somewhat unclear (I had to look up incrementAndGet to make sure it does what I expect). It would be great to move this into a `limitByCount(Predicate, int)` method.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionMultiPartitionTest.java
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionMultiPartitionTest.java:86]: 🔧 I understand that commands appear to be part of the lifecycle, but I'd prefer asserting events only. Events are what really happened, commands are just our implementation to get there. Commands can also be rejected.

**Review by korthout — APPROVED** [review]
Awesome work @remcowesterhoud 👏 

👍 Well split up and described!

💭 It's beautiful to see command distribution in action 🥲 

🔧 Please have a look at my suggestion, I especially want to point out suggestions 2 and 3 as something more important. But it does not have to block merging this

LGTM 👍

File: engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionProcessor.java
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionProcessor.java:68]: Would you consider renaming to `tryDeleteResources` to be enough? I could of course make it return an `Either` but this comes with a tradeoff of duplicating the error handling, which I personally am not a fan of.

File: test-util/src/main/java/io/camunda/zeebe/test/util/stream/StreamWrapper.java
Comment (korthout) [inline test-util/src/main/java/io/camunda/zeebe/test/util/stream/StreamWrapper.java:74]: I was thinking a bit about this, because `count`'s result is unused, and `peek` is used outside of [its intended scope](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#peek-java.util.function.Consumer-).

>This method exists mainly to support debugging, where you want to see the elements as they flow past a certain point in a pipeline

❓ What do you think about this alternative to inclusive take while?
```java
final var index = new AtomicInteger();
final var hit = new AtomicBoolean();
return supply(
    wrappedStream.filter(
        e -> {
          if (hit.get()) {
            return false;
          }

          hit.set(predicate.test(e) && index.getAndIncrement() < count);
          // even when the predicate matches we still need to include the current element
          return true;
        }));
```

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/distribution/CommandDistributionRecord.java
Comment (korthout) [inline protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/distribution/CommandDistributionRecord.java:119]: 🔧 Reading these changes makes me think we should also change the name of `recordValueBuffer` and `concreteRecordValue` as well as the usage of record value in the comments.

**Review by korthout — APPROVED** [review]
Nice work @remcowesterhoud 👏 

🔧 Please consider my comments before merging

LGTM 👍

File: engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionProcessor.java
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionProcessor.java]: ❓ What should happen when the `tryDeleteResources` fails for a distributed command?

**Comment by remcowesterhoud** [comment]
😮 I got a review before requesting it. I'm not even done yet 😛

File: engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionProcessor.java
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/resource/ResourceDeletionProcessor.java]: We should reject the command. If we don't know the resource there is no need to delete it.
In theory this should only happen when the command is distributed twice. I suppose that's where the idempotency comes in. We must be able to handle this command. We do it here by rejecting it.

File: test-util/src/main/java/io/camunda/zeebe/test/util/stream/StreamWrapper.java
Comment (remcowesterhoud) [reply test-util/src/main/java/io/camunda/zeebe/test/util/stream/StreamWrapper.java:74]: I understand why you would say this. I also struggled a bit with different solutions.

Problem is, our wrapped stream wait by default for 5 seconds, so just using this filter isn't sufficient. We need to have some limiting factor to short-circuit the stream, like we do in all our tests. 

The problem here is what do we limit it on. We don't know how many records are expected to be returned. This is why the "dirty" peek workaround is used. We can limit the stream once we've filtered `count` amount of elements. The peek makes sure we get all the records before this as well.
`limit` on it's own doesn't work as it's an intermediate operation of the stream. We need to end on a terminal operation, which is why I use `count` here.

Without all this all tests using the `limit` will run >5 seconds.

**Comment by remcowesterhoud** [comment]
Now it's ready 🙂

**Review by korthout — APPROVED** [review]

**Comment by remcowesterhoud** [comment]
bors merge

**Comment by remcowesterhoud** [comment]
bors retry

**Pull request merged**
